widget/slider.go

```diff
@@ -0,0 +1,289 @@
+package widget
+
+import (
+	"image/color"
+	"math"
+
+	"fyne.io/fyne"
+	"fyne.io/fyne/canvas"
+	"fyne.io/fyne/theme"
+)
+
+const (
+	standardScale = 6
+	minLongSide   = 50
+)
+
+var (
+	_ fyne.Draggable = (*Slider)(nil)
+)
+
+// Slider if a widget that can slide between two fixed values.
+type Slider struct {
+	baseWidget
+
+	Value float64
+	Min   float64
+	Max   float64
+	Step  float64
+
+	Vertical  bool
+	OnChanged func()
+}
+
+// Resize sets a new size for a widget.
+func (s *Slider) Resize(size fyne.Size) {
+	s.resize(size, s)
+}
+
+// MinSize returns the smallest size this widget can shrink to
+func (s *Slider) MinSize() fyne.Size {
+	return s.minSize(s)
+}
+
+// Show this widget, if it was previously hidden
+func (s *Slider) Show() {
+	s.show(s)
+}
+
+// Hide this widget, if it was previously visible
+func (s *Slider) Hide() {
+	s.hide(s)
+}
+
+// Move the widget to a new position, relative to its parent.
+func (s *Slider) Move(pos fyne.Position) {
+	s.move(pos, s)
+}
+
+// DragEnd function.
+func (s *Slider) DragEnd() {
+}
+
+// Dragged function.
+func (s *Slider) Dragged(e *fyne.DragEvent) {
+	ok, pos, max := s.wasSliderEvent(&(e.PointEvent))
+
+	if ok {
+		// clamp the position for drags that go out of bounds
+		if pos > max {
+			pos = max
+		} else if pos < 0 {
+			pos = 0
+		}
+
+		s.fireTrigger(pos, max)
+	}
+}
+
+func (s *Slider) wasSliderEvent(e *fyne.PointEvent) (
+	ok bool, pos, max int) {
+
+	render := Renderer(s).(*sliderRenderer)
+
+	tp := render.thumb.Position()
+	ts := render.thumb.Size()
+
+	pad := theme.Padding()
+
+	x := e.Position.X
+	y := e.Position.Y
+
+	if s.Vertical {
+		if x > (tp.X-pad) && x < (tp.X+ts.Width+pad) {
+			// if the cursor was inside the slider area
+			return true, y, render.track.Size().Height
+		}
+	} else {
+		if y > (tp.Y-pad) && y < (tp.Y+ts.Height+pad) {
+			// if the cursor was inside the slider area
+			return true, x, render.track.Size().Width
+		}
+	}
+	return false, 0, 0
+}
+
+func (s *Slider) fireTrigger(pos, max int) {
+	// update value
+	s.updateValue(float64(pos) / float64(max))
+	Refresh(s)
+
+	if s.OnChanged != nil {
+		s.OnChanged()
+	}
+}
+
+func (s *Slider) updateValue(ratio float64) {
+	if s.Vertical {
+		ratio = 1 - ratio
+	}
+	v := s.Min + ratio*(s.Max-s.Min)
+
+	i := -(math.Log10(s.Step))
+	p := math.Pow(10, i)
+
+	// hack to deal with asymptotic effect for decimal increments
```

Comment:
> it suggests that there is a better way to do something.

Indeed that is what it suggests, and thanks to your insight I found it. Furthermore, I found an edge case that happens in reverse when the min is negative.

```go
if v == s.Max {
	s.Value = s.Max
} else if v == s.Min {
	s.Value = s.Min
} else {
	s.Value = float64(int(v*p)) / p
}
```

Reply:
Don't know the code but to me that looks like those should be >= and <= checks

Reply:
Typically, I would say yes the should be because they are float values, however those are clamped in a check earlier in the code. Having said that it wouldn't break anything to change it over.

Reply:
yes, it's probably clearer with the extra cautious checks